From 0e00381c5967c8f2e4540394a012eb358ac385b3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=D9=85=D9=87=D8=AF=D9=8A=20=D8=B4=D9=8A=D9=86=D9=88=D9=86?=
 =?UTF-8?q?=20=28Mehdi=20Chinoune=29?= <mehdi.chinoune@hotmail.com>
Date: Mon, 26 Feb 2024 18:26:36 +0100
Subject: [PATCH] Generate and Install CMake configuration files

Help to find flint with cmake `find_package(flint)` and link to it directly `target_ling_libraries(target flint::flint)`
---
 CMakeLists.txt       | 48 ++++++++++++++++++++++++++++++++++++--------
 flintConfig.cmake.in | 18 +++++++++++++++++
 2 files changed, 58 insertions(+), 8 deletions(-)
 create mode 100644 flintConfig.cmake.in

diff --git a/CMakeLists.txt b/CMakeLists.txt
index c34047ed4f..c938bd7dcd 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -20,6 +20,7 @@ include(CheckCCompilerFlag)
 include(CheckCSourceRuns)
 include(CheckIPOSupported)
 include(FindPkgConfig)
+include(CMakePackageConfigHelpers)
 
 # Source of truth for project version
 file(STRINGS VERSION FLINT_VERSION_FULL)
@@ -72,8 +73,10 @@ endif()
 
 # Find dependencies
 find_package(PkgConfig REQUIRED)
-pkg_check_modules(GMP REQUIRED IMPORTED_TARGET gmp>=6.2.1)
-pkg_check_modules(MPFR REQUIRED IMPORTED_TARGET mpfr>=4.1.0)
+set(GMP_MIN_VERSION 6.2.1)
+set(MPFR_MIN_VERSION 4.1.0)
+pkg_check_modules(GMP REQUIRED IMPORTED_TARGET gmp>=${GMP_MIN_VERSION})
+pkg_check_modules(MPFR REQUIRED IMPORTED_TARGET mpfr>=${MPFR_MIN_VERSION})
 if (WITH_NTL)
     find_package(NTL REQUIRED)
 endif()
@@ -357,9 +360,13 @@ endif()
 
 # Include directories
 
-target_include_directories(flint PUBLIC 
-    ${CMAKE_CURRENT_SOURCE_DIR}/src ${GMP_INCLUDE_DIRS} ${MPFR_INCLUDE_DIRS}
-    ${CMAKE_CURRENT_BINARY_DIR} ${PThreads_INCLUDE_DIRS})
+target_include_directories(flint PUBLIC
+    "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src;${CMAKE_CURRENT_BINARY_DIR}>"
+    "$<INSTALL_INTERFACE:include/flint>"
+    ${GMP_INCLUDE_DIRS}
+    ${MPFR_INCLUDE_DIRS}
+    ${PThreads_INCLUDE_DIRS}
+)
 
 if(FLINT_USES_BLAS)
     target_include_directories(flint PUBLIC ${CBLAS_INCLUDE_DIRS})
@@ -437,13 +444,38 @@ endif()
 include(GNUInstallDirs)
 
 install(TARGETS flint
-            RUNTIME DESTINATION "${CMAKE_INSTALL_FULL_BINDIR}"
-            ARCHIVE DESTINATION "${CMAKE_INSTALL_FULL_LIBDIR}"
-            LIBRARY DESTINATION "${CMAKE_INSTALL_FULL_LIBDIR}"
+            EXPORT flintTargets
+            RUNTIME DESTINATION "${CMAKE_INSTALL_BINDIR}"
+            ARCHIVE DESTINATION "${CMAKE_INSTALL_LIBDIR}"
+            LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}"
         )
 
 install(FILES ${HEADERS} DESTINATION include/flint)
 
+write_basic_package_version_file(
+  "${CMAKE_CURRENT_BINARY_DIR}/flintConfigVersion.cmake"
+  VERSION "${PROJECT_VERSION}"
+  COMPATIBILITY AnyNewerVersion
+)
+
+configure_package_config_file(${CMAKE_CURRENT_SOURCE_DIR}/flintConfig.cmake.in
+  "${CMAKE_CURRENT_BINARY_DIR}/flintConfig.cmake"
+  INSTALL_DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/flint
+)
+
+install(FILES
+        "${CMAKE_CURRENT_BINARY_DIR}/flintConfig.cmake"
+        "${CMAKE_CURRENT_BINARY_DIR}/flintConfigVersion.cmake"
+        "CMake/FindCBLAS.cmake"
+        DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/flint
+)
+
+install(EXPORT flintTargets
+        FILE flintTargets.cmake
+        NAMESPACE flint::
+        DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/flint
+)
+
 set_target_properties(flint
     PROPERTIES
     ARCHIVE_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/lib"
diff --git a/flintConfig.cmake.in b/flintConfig.cmake.in
new file mode 100644
index 0000000000..24336719ac
--- /dev/null
+++ b/flintConfig.cmake.in
@@ -0,0 +1,18 @@
+@PACKAGE_INIT@
+
+find_package(PkgConfig REQUIRED)
+pkg_check_modules(GMP REQUIRED IMPORTED_TARGET gmp>=@GMP_MIN_VERSION@)
+pkg_check_modules(MPFR REQUIRED IMPORTED_TARGET mpfr>=@MPFR_MIN_VERSION@)
+
+if(@CBLAS_FOUND@)
+  find_package(CBLAS REQUIRED)
+endif()
+
+if(NOT MSVC)
+  list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}")
+  find_package(Threads REQUIRED)
+endif()
+
+include("${CMAKE_CURRENT_LIST_DIR}/flintTargets.cmake")
+
+check_required_components(flint)
